design/m1_push_to_file_design.md

```diff
@@ -0,0 +1,1247 @@
+| 10 | File permissions error | You are using SP in "file" mode | Secrets Provider does not have permission to write a secrets file to the specified volume and mount path. | Error: Secrets Provider does not have permission to write the secrets file "{value}". Please check the volume mount permissions on the Secrets Provider container and try again.<br><br>SP fails | | |
+| 11 | Invalid file path provided | You are using SP in "file" mode and have provided a value for the conjur.org/secret-file-path annotation | The provided file path is not a valid file path. | Error: You have provided an invalid file path "{value}" for the "{value}" secret group.<br><br>SP fails to start | | |
+| 12 | File conflict between two secret groups | You are using SP in "file" mode and have defined multiple secret groups | One or more secret groups have he same ouput file. | Error: You have provided conflicting file paths; secret groups "{value}", "{value}", ... all have output file "{value}".<br><br>SP fails to start | | |
+
+#### Integration / E2E tests
+
+E2E tests are expensive. We should try to limit their numbers. There are cheaper ways to validate behavior. For the push to file functionality it will likely be possible to validate the happy path in one fell swoop by using configuration tht captures all the supported annotations and template types.
+
+- [] Happy path, Secret Provider runs as an init container in file mode and successfully injects files. This single test can cover all supported annotations and template types.
+
+The sad path is where things tend to become difficult, because there's a myriad of ways in which failure can occur. A possible way to reduce the need to E2E test failures is to establish a nexus point for all failures in the code so that an error can be validated at the unit/functional testing level.
+
+- [] Sad path, Secret Provider runs as an init container in file mode and fails, thereby preventing the deployment from proceeding. TBC it's possible we don't need this at all.
+
+#### Security testing
+
+Security testing will include:
+
+- Automated Vulnerability scans
+
+####  Performance testing
+
+As described in the [Performance](#performance) section above, the performance
+of the Secrets Provider for Push-to-File mode will be manually measured using
+the following metrics:
+
+- Latency (processing time) for parsing Pod annotations
+- Latency for generating secrets requests and processing responses
+- Latency for writing retrieved secrets to files
+
+_**QUESTION: Do we need to measure CPU and memory footprints?**_
+
+The measurements will be made using a representative configuration in terms
+of the number of secrets groups and the number of secrets per secrets group.
+Details are TBD, but a representative configuration might look something like
+this:
+
+- 5 secrets groups
```

Comment:
@alexkalish This is good information. I added this to the `Project Scope and Limitations` section, and mentioned that the number of Conjur secrets used in performance testing:
```
      5 Conjur secrets x 5 secrets groups = 25 Conjur secrets total
```
represents about half of the Secrets Provider's scaling limit.

Reply:
@andytinkham , We'll only have one Secrets Provider init container per application Pod, so technically speaking there won't be multiple Secrets Providers writing to the secrets files.

However, your point is valid that we need to account for the potential for multiple "file writers" (with one file writer per configured secrets group) writing to the same secrets file.

For the short term, I'm assuming that we can consider a configuration that uses the same destination secrets file name for multiple groups to be a (fatal) misconfiguration. We'll add a validation check for this after all annotations have been parsed.

I've also added a note that **in the future**, we may consider adding concurrency protection mechanisms that would allow for multiple file writers (i.e. multiple secrets groups) to append secrets settings to a common secrets file. This is technically possible to implement, but I don't know if there are a lot of use cases where this would be required.

Reply:
My table of errors in the Feature Doc calls out this case as an error.  We should fail early if two secret groups specify the same file path.